components/script/dom/htmldocument.rs

```diff
@@ -0,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::ptr::NonNull;
+
+use dom_struct::dom_struct;
+use js::jsapi::JSObject;
+
+use crate::dom::bindings::codegen::Bindings::HTMLDocumentBinding::HTMLDocumentMethods;
+use crate::dom::bindings::error::{Error, Fallible};
+use crate::dom::bindings::reflector::Reflector;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::location::Location;
+use crate::script_runtime::JSContext as SafeJSContext;
+
+// https://html.spec.whatwg.org/multipage/#htmldocument
+#[dom_struct]
+pub struct HTMLDocument {
+    reflector_: Reflector,
+}
+
+impl HTMLDocumentMethods for HTMLDocument {
+    // https://html.spec.whatwg.org/multipage/#htmldocument
+    fn GetLocation(&self) -> Option<DomRoot<Location>> {
+        None
+    }
+
+    // https://html.spec.whatwg.org/multipage/#htmldocument
+    fn SupportedPropertyNames(&self) -> Vec<DOMString> {
+        Vec::new()
+    }
+
+    // https://html.spec.whatwg.org/multipage/#htmldocument
+    fn NamedGetter(
+        &self,
+        _cx: SafeJSContext,
+        _name: DOMString,
+    ) -> Fallible<Option<NonNull<JSObject>>> {
+        Err(Error::NotFound)
+    }
```

Comment:
Could these not be implemented?

Reply:
Is it not enough to simply ensure that window has an HTMLDocument getter that does the same thing as `fn Document(...)` in `window.rs`?

Reply:
Reading through https://github.com/whatwg/html/issues/4792 this seems to be the situation:

 - The specification says that `HTMLDocument` should be an alias for `Document` ie we should not need to implement a separate document interface. The `Document` interface implements the named getter (Servo gained support for this unfortunate part of the web platform last year).
 - No browsers do this though. Instead FireFox has `window.document` implement `HTMLDocument`, but not for XML documents. This means that `responseXML` and other XML pages do not have a named getter, which seems to big the biggest compatability risk.

This change is quite strange though. It implements `HTMLDocument`, but `window.document` is still not an HTMLDocument. If we want to follow Firefox, it seems like the named getter should move to `HTMLDocument` and `HTMLDocument` should be exposed at `window.document` when the document is HTML (but not XML).